api/v1alpha4/cluster_types.go

```diff
@@ -60,6 +60,69 @@ type ClusterSpec struct {
 	// for provisioning infrastructure for a cluster in said provider.
 	// +optional
 	InfrastructureRef *corev1.ObjectReference `json:"infrastructureRef,omitempty"`
+
+	// This encapsulates the topology for the cluster.
+	// NOTE: This feature is alpha; it is required to enable the ClusterManagedTopologies
+	// feature gate flag to activate managed topologies support.
+	// +optional
+	Topology *Topology `json:"topology,omitempty"`
+}
+
+// Topology encapsulates the information of the managed resources.
+type Topology struct {
+	// The name of the ClusterClass object to create the topology.
+	Class string `json:"class,omitempty"`
+
+	// The kubernetes version of the cluster.
+	Version string `json:"version"`
+
+	// RolloutAfter performs a rollout of the entire cluster one component at a time,
+	// control plane first and then machine deployments.
+	// +optional
+	RolloutAfter *metav1.Time `json:"rolloutAfter,omitempty"`
+
+	// The information for the Control plane of the cluster.
+	ControlPlane ControlPlaneTopology `json:"controlPlane"`
+
+	// Worker encapsulates the different constructs that form the worker nodes
+	// for the cluster.
+	// +optional
+	Worker *WorkerTopology `json:"worker,omitempty"`
+}
+
+// ControlPlaneTopology specifies the parameters for the control plane nodes in the cluster.
+type ControlPlaneTopology struct {
+	Metadata ObjectMeta `json:"metadata,omitempty"`
+
+	// The number of control plane nodes.
+	Replicas int `json:"replicas"`
+}
+
+// WorkerTopology represents the different sets of worker nodes in the cluster.
+type WorkerTopology struct {
+	// Deployments is a list of machine deployment in the cluster.
+	MachineDeployments []MachineDeploymentTopology `json:"machineDeployments,omitempty"`
+}
+
+// MachineDeploymentTopology specifies the different parameters for a set of worker nodes in the topology.
+// This set of nodes is managed by a MachineDeployment object whose lifecycle is managed by the Cluster controller.
+type MachineDeploymentTopology struct {
+	Metadata ObjectMeta `json:"metadata,omitempty"`
```

Comment:
what is the purpose of metadata here?

Reply:
Metadata propagates labels an annotations to objects, it's used like we use it today in MachineDeployments templates